contracts/xastro_token/src/contract.rs

```diff
@@ -0,0 +1,700 @@
+use cosmwasm_std::{
+    attr, entry_point, to_binary, Binary, Deps, DepsMut, Env, MessageInfo, Order, Response,
+    StdError, StdResult, Uint128,
+};
+use cw20::{AllAccountsResponse, BalanceResponse, Cw20Coin, Cw20ReceiveMsg};
+use cw20_base::allowances::{
+    deduct_allowance, execute_decrease_allowance, execute_increase_allowance, query_allowance,
+};
+
+use crate::state::{capture_total_supply_history, get_total_supply_at, BALANCES};
+use astroport::asset::addr_validate_to_lower;
+use cw2::set_contract_version;
+use cw20_base::contract::{
+    execute_update_marketing, execute_upload_logo, query_download_logo, query_marketing_info,
+    query_minter, query_token_info,
+};
+use cw20_base::enumerable::query_all_allowances;
+use cw20_base::msg::ExecuteMsg;
+use cw20_base::state::{MinterData, TokenInfo, TOKEN_INFO};
+use cw20_base::ContractError;
+use cw_storage_plus::Bound;
+
+use astroport::xastro_token::{InstantiateMsg, QueryMsg};
+
+/// Contract name that is used for migration.
+const CONTRACT_NAME: &str = "astroport-xastro-token";
+/// Contract version that is used for migration.
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+// settings for pagination
+const MAX_LIMIT: u32 = 30;
+const DEFAULT_LIMIT: u32 = 10;
+
+/// ## Description
+/// Creates a new contract with the specified parameters in the [`InstantiateMsg`].
+/// Returns the default object of type [`Response`] if the operation was successful,
+/// or a [`ContractError`] if the contract was not created.
+/// ## Params
+/// * **deps** is the object of type [`DepsMut`].
+///
+/// * **_env** is the object of type [`Env`].
+///
+/// * **_info** is the object of type [`MessageInfo`].
+/// * **msg** is a message of type [`InstantiateMsg`] which contains the basic settings for creating a contract.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    mut deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> StdResult<Response> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Check valid token info
+    msg.validate()?;
+
+    // Create initial accounts
+    let total_supply = create_accounts(&mut deps, &env, msg.initial_balances.as_slice())?;
+
+    capture_total_supply_history(deps.storage, &env, total_supply)?;
```

Comment:
Let's not call it if total_supply is zero

Reply:
Ok, done. Thanks